### PR TITLE
vector testing support - restructured branch for easier PR reviews

### PR DESCRIPTION
**Pull request by jshook**
This includes several commits which have been generally grouped thematically. It will make review easier if you select which commit you want to look at individually during review.

### CODE REVIEW

File: adapters-api/src/main/java/io/nosqlbench/adapters/api/evalcontext/GroovyCycleFunction.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+import org.apache.logging.log4j.Logger;
+import org.codehaus.groovy.control.CompilerConfiguration;
+import org.codehaus.groovy.control.customizers.ImportCustomizer;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public class GroovyCycleFunction<T> implements CycleFunction<T> {
+    private final static Logger logger = LogManager.getLogger(GroovyBooleanCycleFunction.class);
```
Comment (jeffbanks) [inline adapters-api/src/main/java/io/nosqlbench/adapters/api/evalcontext/GroovyCycleFunction.java:37]: Logging class should be corrected.

File: adapters-api/src/test/java/io/nosqlbench/adapters/api/evalcontext/GroovyBooleanCycleFunctionTest.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    @Test
+    public void testUncaughtException() {
+        ParsedTemplateString parsedTemplate = new ParsedTemplateString("""
+            this_is_a_syntax_error
+            """,
+            Map.of("numbername", "NumberNameToString()")
+        );
+        List<String> imports = List.of();
+        GroovyCycleFunction function = new GroovyBooleanCycleFunction("test2", parsedTemplate, imports);
+        System.out.println(function);
```
Comment (jeffbanks) [inline adapters-api/src/test/java/io/nosqlbench/adapters/api/evalcontext/GroovyBooleanCycleFunctionTest.java:53]: Consider removing sysout.

**Review by jeffbanks — CHANGES_REQUESTED** [review]
Overall looks good.  I will pull in the `evalcontext` additions (Groovy*) to understand that a bit more.  I suspect we will need some complementing docs.  Just a few suggestions for changes/corrections.

File: adapters-api/src/main/java/io/nosqlbench/adapters/api/activityconfig/rawyaml/RawOpDef.java [hunk 3]
```diff
@@ -57,9 +65,19 @@ public void setFieldsByReflection(Map<String, Object> map) {
...
-            setOp(op);
-        } else if (found.size() > 1) {
+            String keyName = found.iterator().next();
+            Object op = map.remove(keyName);
+            if (op instanceof CharSequence s) {
+                if (!keyName.equals("stmt")) {
+                    logger.warn("Used implied stmt field under name '" + keyName + "'. You can just use 'stmt: ... "+ s +"' or the equivalent to avoid this warning.");
+                }
+                map.put("stmt",s.toString());
+//                setOp(new LinkedHashMap<String,Object>(Map.of("stmt",s.toString())));
```
[the remark this answers is not in the document]
Comment (jshook) [reply adapters-api/src/main/java/io/nosqlbench/adapters/api/activityconfig/rawyaml/RawOpDef.java:75]: replacing the logic, but the comment is cruft

File: adapters-api/src/main/java/io/nosqlbench/adapters/api/activityconfig/rawyaml/RawOpDef.java [hunk 4]
```diff
@@ -57,9 +65,19 @@ public void setFieldsByReflection(Map<String, Object> map) {
...
         }
         if (found.size() == 1) {
-            Object op = map.remove(found.iterator().next());
-            setOp(op);
-        } else if (found.size() > 1) {
+            String keyName = found.iterator().next();
+            Object op = map.remove(keyName);
+            if (op instanceof CharSequence s) {
+                if (!keyName.equals("stmt")) {
+                    logger.warn("Used implied stmt field under name '" + keyName + "'. You can just use 'stmt: ... "+ s +"' or the equivalent to avoid this warning.");
```
[the remark this answers is not in the document]
Comment (jshook) [reply adapters-api/src/main/java/io/nosqlbench/adapters/api/activityconfig/rawyaml/RawOpDef.java:72]: agreed in general, this is a low-frequency high-visibility line, so not going to force it for now

File: adapters-api/src/main/java/io/nosqlbench/adapters/api/evalcontext/GroovyCycleFunction.java (hunk 1 above)
Comment (jshook) [reply adapters-api/src/main/java/io/nosqlbench/adapters/api/evalcontext/GroovyCycleFunction.java:37]: good catch

File: adapters-api/src/test/java/io/nosqlbench/adapters/api/evalcontext/GroovyBooleanCycleFunctionTest.java (hunk 2 above)
Comment (jshook) [reply adapters-api/src/test/java/io/nosqlbench/adapters/api/evalcontext/GroovyBooleanCycleFunctionTest.java:53]: good catch

File: adapter-pinecone/src/main/java/io/nosqlbench/adapter/pinecone/ops/PineconeDeleteOp.java [hunk 5]
```diff
@@ -41,8 +41,9 @@ public PineconeDeleteOp(PineconeConnection connection, DeleteRequest request) {
     }
 
     @Override
-    public void run() {
+    public Object apply(long value) {
         DeleteResponse response = connection.getBlockingStub().delete(request);
         logger.debug("Pinecone delete request successful: " + response.toString());
+        return response;
```
[the remark this answers is not in the document]
Comment (jshook) [reply adapter-pinecone/src/main/java/io/nosqlbench/adapter/pinecone/ops/PineconeDeleteOp.java:47]: good catch

File: adapter-pinecone/src/main/resources/activities/pinecone-verify.yaml [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+  default:
+    mixed: run driver=pinecone cycles=1000 apiKey=2f55b2f0-670f-4c51-9073-4d37142b761a projectName=a850334 environment=us-east-1-aws tags='block:main-.*'
+
+bindings:
+  id: Mod(<<keycount:1000000000>>); ToString() -> String
+  vector_value: CircleVectors(100000, "io.nosqlbench.virtdata.library.basics.shared.vectors.algorithms.GoldenAngle")
+  state: StateCodes()
+
+blocks:
+#  main-write:
```
[the remark this answers is not in the document]
Comment (jshook) [reply adapter-pinecone/src/main/resources/activities/pinecone-verify.yaml:11]: leaving as-is for now

File: adapter-cqld4/src/main/java/io/nosqlbench/datamappers/functions/geometry/NormalizeCqlVector.java [hunk 7]
```diff
@@ -33,37 +33,28 @@
...
-            list.forEach(o -> doubles.add((Double) o));
-            for (Double dv : doubles) {
-                builder.add(dv);
-            }
+    public CqlVector apply(CqlVector cqlVector) {
+        double[] vals = new double[cqlVector.size()];
+        double accumulator= 0.0d;
+        for (int i = 0; i < vals.length; i++) {
+            vals[i]=cqlVector.get(i).doubleValue();
+            accumulator+=vals[i]*vals[i];
```
[the remark this answers is not in the document]
Comment (jshook) [reply adapter-cqld4/src/main/java/io/nosqlbench/datamappers/functions/geometry/NormalizeCqlVector.java:46]: this is likely a bug

[the remark this answers is not in the document]
Comment (jshook) [reply adapter-cqld4/src/main/java/io/nosqlbench/datamappers/functions/geometry/NormalizeCqlVector.java:46]: The streaming call below replaces the accumulator

**Review by MarkWolters — APPROVED** [review]
Reviewed online w/Shooky, looks good!

**Pull request merged**